File: Python/cnpj.py
import re
from typing import Union, List, Optional


class CNPJ:
    """
    Classe utilitária para validação e formatação de CNPJ brasileiro.
    """

    _regex_cnpj = re.compile(r"^\d{2}\.\d{3}\.\d{3}/\d{4}-\d{2}$")

    @staticmethod
    def is_valid(cnpj: Union[str, int, List[int], None]) -> bool:
        """
        Verifica se o CNPJ fornecido é válido.
        Aceita strings com ou sem máscara, inteiros ou listas de inteiros.
        """
        if cnpj is None:
            return False

        if not isinstance(cnpj, (str, int, list)):
            return False

        if isinstance(cnpj, str):
            digits_only = re.fullmatch(r"\d{14}", cnpj) is not None
            valid_format = CNPJ._regex_cnpj.fullmatch(cnpj) is not None
            if not digits_only and not valid_format:
                return False

        numbers = CNPJ._extract_numbers(cnpj)
        if len(numbers) != 14 or len(set(numbers)) == 1:
            return False

        first_digit = CNPJ._calculate_digit(12, numbers)
        if first_digit != numbers[12]:
            return False

        second_digit = CNPJ._calculate_digit(13, numbers)
        return second_digit == numbers[13]

    @staticmethod
    def format(cnpj: Union[str, int, List[int]]) -> str:
        """
        Retorna o CNPJ formatado como 00.000.000/0000-00.
        Retorna string vazia se o CNPJ não for válido.
        """
        if not CNPJ.is_valid(cnpj):
            return ""

        numbers = CNPJ._extract_numbers(cnpj)
        digits = "".join(map(str, numbers))
        return (
            f"{digits[:2]}.{digits[2:5]}.{digits[5:8]}/{digits[8:12]}-{digits[12:14]}"
        )

    @staticmethod
    def _extract_numbers(value: Union[str, int, List[int]]) -> List[int]:
        """
        Extrai apenas os dígitos numéricos do valor fornecido.
        """
        return [int(d) for d in re.sub(r"\D", "", str(value))]

    @staticmethod
    def _calculate_digit(x: int, numbers: List[int]) -> int:
        """
        Calcula o dígito verificador com base nos primeiros `x` números.
        """
        slice_ = numbers[:x]
        factor = x - 7
        total = 0

        for i in range(x, 0, -1):
            n = slice_[x - i]
            total += n * factor
            factor -= 1
            if factor < 2:
                factor = 9

        result = 11 - (total % 11)
        return 0 if result > 9 else result

File: Swift/CNPJ.swift
import Foundation

class CNPJ {
    
    private static let regexCNPJ = try! NSRegularExpression(pattern: #"^\d{2}\.\d{3}\.\d{3}/\d{4}-\d{2}$"#)

    static func isValid(_ cnpj: Any?) -> Bool {
        guard let cnpj = cnpj else { return false }
        
        let stringValue: String
        if let s = cnpj as? String {
            stringValue = s
            let digitsOnly = stringValue.range(of: #"^\d{14}$"#, options: .regularExpression) != nil
            let matchesFormat = regexCNPJ.firstMatch(in: stringValue, options: [], range: NSRange(location: 0, length: stringValue.utf16.count)) != nil
            if !digitsOnly && !matchesFormat {
                return false
            }
        } else if let number = cnpj as? Int {
            stringValue = String(format: "%014d", number)
        } else if let list = cnpj as? [Int], list.count == 14 {
            stringValue = list.map(String.init).joined()
        } else {
            return false
        }
        
        let numbers = extractNumbers(from: stringValue)
        guard numbers.count == 14 else { return false }
        if Set(numbers).count == 1 { return false }

        let firstDigit = calculateDigit(x: 12, numbers: numbers)
        if firstDigit != numbers[12] { return false }

        let secondDigit = calculateDigit(x: 13, numbers: numbers)
        return secondDigit == numbers[13]
    }

    static func format(_ cnpj: Any?) -> String {
        guard isValid(cnpj) else { return "" }
        let numbers = extractNumbers(from: cnpj!)
        let digits = numbers.map(String.init).joined()
        
        return "\(digits.prefix(2)).\(digits[2..<5]).\(digits[5..<8])/\(digits[8..<12])-\(digits.suffix(2))"
    }

    private static func extractNumbers(from value: Any) -> [Int] {
        let str = String(describing: value)
        return str.filter { $0.isNumber }.compactMap { Int(String($0)) }
    }

    private static func calculateDigit(x: Int, numbers: [Int]) -> Int {
        let slice = numbers.prefix(x)
        var factor = x - 7
        var sum = 0
        
        for n in slice {
            sum += n * factor
            factor -= 1
            if factor < 2 { factor = 9 }
        }
        
        let result = 11 - (sum % 11)
        return result > 9 ? 0 : result
    }
}


